### PR TITLE
Adopt InternMap.

### LINKED ISSUE #184: Adopt InternMap for d3.group et al.?
Should d3.group, d3.rollup, et al. return an [InternMap](https://observablehq.com/@mbostock/internmap) instead of a regular Map? That would allow you to group by date easily. I think we can even consider it backwards-compatible since the behavior is the same for primitives and for objects that do not implement *object*.valueOf (which is most things besides Date and the rarely-used Number, String, and Boolean).

### PR DESCRIPTION
**Pull request by mbostock**
Fixes #184.

Playground: https://observablehq.com/d/d4c5f6ad343866b9

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -50,3 +50,4 @@ export {default as intersection} from "./intersection.js";
 export {default as subset} from "./subset.js";
 export {default as superset} from "./superset.js";
 export {default as union} from "./union.js";
+export {InternMap, InternSet} from "internmap";
```
Comment (curran) [inline src/index.js:53]: Why export these here?

Is the intention for these to end up as `d3.InternMap` and `d3.InternSet` (via [`export * from "d3-array";`](https://github.com/d3/d3/blob/master/index.js#L2))?

Comment (mbostock) [reply src/index.js:53]: Yes.

**Comment by Fil** [comment]
Documentation at https://observablehq.com/@d3/d3-group will need an update when this lands. (I've also added a section on d3.index, if you want to proofread it.)

Note that the paragraphs about [key equality](https://observablehq.com/@d3/d3-group#key_equality) are defensive, and we'll need to mention the version number which solves this.

It's worth noting that while InternMap can index on an array, by passing a key function, there is no way to pass a key function to d3.group (and rollup and index). When the key is on the first level it's easy though to upgrade to an InternMap:

```js
const g = new d3.InternMap(
  d3.group(
    [[["foo", "bar"], 1], [["foo", "baz"], 2], [["goo", "bee"], 3]],
    d => d[0]
  ),
  JSON.stringify
);
g.get(["foo", "bar"]); // 1
```

**Comment by Fil** [comment]
I've added documentation in #186.

**Comment by mbostock** [comment]
I thought about exposing the configurable key function through d3.group et al. but that feels like overkill to start. (I mainly think configurable keys are useful for compound keys; compound keys aren’t typically needed with d3.group because it’s already nested, and furthermore JSON.stringify is slow.) If we do need to expose it in the future, it could be something like `d3.group.key(JSON.stringify)(data, d => d.foo)`, or `d3.groupKey` etc.

Thanks for writing docs! I’ll check ’em out now.

**Pull request merged**